src/RawTransaction.php

```diff
@@ -1208,29 +1208,35 @@ public static function extract_input_signatures_p2sh($input, $message_hash, $scr
      */
     public static function encode_signature(Signature $signature)
     {
-
         // Pad r and s to 64 characters.
-        $rh = str_pad(BitcoinLib::hex_encode($signature->getR()), 64, '0', STR_PAD_LEFT);
-        $sh = str_pad(BitcoinLib::hex_encode($signature->getS()), 64, '0', STR_PAD_LEFT);
-
-        // Check if the first byte of each has its highest bit set,
-        $t1 = unpack("H*", (pack('H*', substr($rh, 0, 2)) & pack('H*', '80')));
-        $t2 = unpack("H*", (pack('H*', substr($sh, 0, 2)) & pack('H*', '80')));
-        // if so, the result != 00, and must be padded.
-        $r = ($t1[1] !== '00') ? '00' . $rh : $rh;
-        $s = ($t2[1] !== '00') ? '00' . $sh : $sh;
-
-        // Create the signature.
-        $der_sig = '30'
-            . self::_dec_to_bytes((4 + ((strlen($r) + strlen($s)) / 2)), 1) //((strlen($r)+strlen($s)+16)/2),1)
-            . '02'
-            . self::_dec_to_bytes(strlen($r) / 2, 1)
-            . $r
-            . '02'
-            . self::_dec_to_bytes(strlen($s) / 2, 1)
-            . $s
-            . '01';
+        $rBin = pack("H*", BitcoinLib::hex_encode($signature->getR()));
+        $sBin = pack("H*", BitcoinLib::hex_encode($signature->getS()));
+
+        // Pad R and S if their highest bit is flipped, ie,
+        // they are negative.
+        $rt = $rBin[0] & pack('H*', '80');
+        if (ord($rt) == 128) {
+            $rBin = pack('H*', '00') . $rBin;
+        }
+
+        $st = $sBin[0] & pack('H*', '80');
+        if (ord($st) == 128) {
+            $sBin = pack('H*', '00') . $sBin;
+        }
+
+        $inner = hex2bin('02')
+            .hex2bin(dechex(strlen($rBin)))
+            .$rBin
+            .hex2bin('02')
+            .hex2bin(dechex(strlen($sBin)))
+            .$sBin;
 
+        $outer = hex2bin('30')
+            .hex2bin(dechex(strlen($inner)))
+            .$inner
+            .hex2bin('01');
```

Comment:
@afk11 @btcdrak I think this 01 at the end should not be here, bitcore and bitcoinjs don't add this and they both see this our signatures as non canonical ...

you also seem to compensate for this suffix in various places, such as doing -3 on the length byte when checking if the sig is canonical, but the length byte should be 1 (length byte R) + len(R) + 1 (length byte S) + len(S).
so the length of the sig should be length byte + 1 (sequence) + 1 (length byte)